README.md
# Continuous Delivery Configuration Server

<img src="https://raw.githubusercontent.com/markusfalk/cd-config-server/main/src/_assets/img/cd-config-server-flow.svg?token=AAKKHMFL3CFGKOX7A4DDZIC76L5IQ" alt="" style="display: block; margin: 50px auto;"/>

This server provides a configuration API to be used within Continuous Delivery environments.

## Concept

The idea behind this software is to provide a global and public service for your apps to receive configuration values from. Configuration can be stored in a git repository that is seperated from the repositories in which your apps reside or synced to your file system. This makes it possible for apps and configurations to have independent lifecyles and both be treated like code.

This service allows for apps to be build once and then put through your continuous delivery pipeline without the need to rebuild them. They can load their configuration at runtime by telling the service what they need.

## How it works

### Application Configuration

A continuous delivery pipeline usually has mutlipe environments in which the app is tested or reviewed.
The configuration repository contains one file per stage/environment of your pipeline.

These configuration files (\*.json) contain all the values your app needs.
If a pipeline has the four stages: development (local), test, staging, production, then you would create the following four files:

```txt
development.json
test.json
staging.json
production.json
```

It currently supports Github, Gitlab or the local file system as a source of your configuration files.

### Independent Lifecyles and Deployments

Here is where the magic happens and where this configuration server is different from others.

To make applications and configurations truly independent and to avoid lock-step releases with your configuration, we need to somehow tell the service which configuration release is compatible with what version of our applications.

For that reason each configuration file contains this one additional and obligatory entry:

```json
{
  "compatibleWithAppVersion": "^2.0.0"
}
```

Here you can define a semantic version range that tells the configuration server if this tag/release of your configuration is comapitble with the application version that is requesting configuration.

Note however, that in case of a breaking change in your app relating to a new configuration entry this will again become a required release with your application. You can release your configuration first by specifing a breaking change. Older apps will not consume the new configuration and as soon as your new app is deployed, the configuration is already there.

### Naming Convention

#### Configuration Repository Name

The application id will be used to find the configuration repository by appending `-config`.

| application id requested in the url | configuration repositoy | url of your configuration repository            |
| ----------------------------------- | ----------------------- | ----------------------------------------------- |
| YourApp                             | YourApp-config          | https://github.com/yournamespace/yourapp-config |
| Another                             | Another-config          | https://gitlab.com/yournamespace/another-config |

Examples

| request url                              | url of your configuration repository                       |
| ---------------------------------------- | ---------------------------------------------------------- |
| /cd-config-server-test/1.0.0/development | https://github.com/markusfalk/cd-config-server-test-config |
| /my-app/1.0.0/development                | https://gitlab.com/markus_falk/my-app-config               |

#### Matching Configuration Key

To match application version and configuration you must provide the `compatibleWithAppVersion` key in all of your configuration files.

```json
{
  "compatibleWithAppVersion": "^1.0.0"
}
```

⚠️ Be aware that this is the version of the application that this specific configuration is compatible with not the version of the configuration itself. They vary because they now have different lifecycles.

⚠️ If the configuration has a breaking change, make sure to deploy the config first to avoid apps in production without propper configuration and be aware of the caching that is configured for configuration output.

In case of a none-breaking change in the configuration file, you would release a new version of the configuration that is still compatible with already deployed app causing for the config to have a new release but with no need to release a new app.

## Setup

### Prerequisits

- Configuration files hosted on filesystem, github or gitlab (including self hosted)
- This server is provided as a docker image
- Your applications need be versioned using semantic versioning

### Options

The server is provided as a docker image and can be started with the following options:

#### Common

| name           | default | description                                                                   |
| -------------- | ------- | ----------------------------------------------------------------------------- |
| CACHE_TTL      | 300     | This service uses an in memory cache and this is how long it lives in seconds |
| CORS_ORIGIN    | \*      | Setup which origins are allowed to access                                     |
| SOURCE\*       |         | 'gitlab' \| 'github' \| 'filesystem'                                          |
| HTTPPORT       | 3000    | What port would you like docker to expose this service to                     |
| RATE_LIMIT_MAX | 5000    | How many requests do you allow to this service per RATE_LIMIT_MS              |
| RATE_LIMIT_MS  | 360000  | Rate limit time range in milliseconds                                         |

#### When using Github

| name             | default | description             |
| ---------------- | ------- | ----------------------- |
| GITHUBPASSWORD\* |         | used for authentication |
| GITHUBUSERNAME\* |         | used for authentication |
| USERAGENT\*      |         | used for authentication |

#### When using Gitlab

| name             | default                   | description                     |
| ---------------- | ------------------------- | ------------------------------- |
| GITLAB_BASE_URL  | https://gitlab.com/api/v4 | address your self hosted gitlab |
| GITLABUSERNAME\* |                           | used to find your project       |

#### When using the file system

There is no additional configuration needed beyond the `SOURCE` but there needs to be a volume mounted to `/app/dist/configfiles`.

When this is done you can move your configuration files into that volume with the following structure:

```txt
- <appid> // identifies the app that is asking for configuration
  |- <configuration release version> // do not confuse with app version. This is the releases of your configuration
    |- <stage>.json // identifies the stage your app wants configuration for
```

Example:

```txt
- my-app
  |- 1.0.0
    |- development.json
    |- test.json
    |- production.json
- another-app
  |- 1.0.0
    |- development.json
    |- test.json
    |- production.json
  |- 1.1.0
    |- development.json
    |- test.json
    |- production.json
  |- 2.0.0
    |- development.json
    |- test.json
    |- production.json
```

## Caveats

If you are using this configuration server publicly you need to make sure that your configurations files do not contain any secrets.

## Migration Guide

- https://github.com/markusfalk/cd-config-server/blob/main/MIGRATIONS.md

## Issues

- https://github.com/markusfalk/cd-config-server/issues

## Changelog

- https://github.com/markusfalk/cd-config-server/blob/main/CHANGELOG.md

## Questions?

If you have any questions reach out to me on twitter: https://twitter.com/markus_falk


CHANGELOG.md
# Changelog

## [3.3.0](https://github.com/markusfalk/cd-config-server/compare/v3.2.0...v3.3.0) (2021-07-31)


### Features

* **github:** improve error messaging ([801a34d](https://github.com/markusfalk/cd-config-server/commit/801a34dcd2bd05cd3dee9eed5af400f2cb51f792))
* upgrade to nestjs 8 ([a717296](https://github.com/markusfalk/cd-config-server/commit/a7172960d1b3cf0557cd74e0a1f459b857b57d71))


### Bug Fixes

* improve error handling ([0364e4f](https://github.com/markusfalk/cd-config-server/commit/0364e4fdf1f6c4a0c835a796382d0e3b52de5271))
* migrate imports ([a0894e1](https://github.com/markusfalk/cd-config-server/commit/a0894e1f14b4f289dfbdaddd673002c0e25ebdde))

## [3.2.0](https://github.com/markusfalk/cd-config-server/compare/v3.1.0...v3.2.0) (2021-05-17)


### Features

* update nest.js ([4ea3a8b](https://github.com/markusfalk/cd-config-server/commit/4ea3a8b3b3f9a2e87abcd49c2999bd5c9323f915))
* **file-system:** improve error handling and logging ([fd07095](https://github.com/markusfalk/cd-config-server/commit/fd07095decf1932068660d1efc101f23016f7205))
* **github:** improve error handling and logging ([4eae0f0](https://github.com/markusfalk/cd-config-server/commit/4eae0f02633b5a23301c0145d64e9424afcdcc83))
* **gitlab:** improve error handling and logging ([1545d84](https://github.com/markusfalk/cd-config-server/commit/1545d84af2815bc4ee93908b3f1c5db29fb0e891))


### Bug Fixes

* **file-system:** use absolute path after refactoring ([00e0881](https://github.com/markusfalk/cd-config-server/commit/00e08810a69e460960fc1f2553acf00285bbe8df))

## [3.1.0](https://github.com/markusfalk/cd-config-server/compare/v3.0.0...v3.1.0) (2021-03-04)


### Features

* add helmet for security ([8dadf7c](https://github.com/markusfalk/cd-config-server/commit/8dadf7c8d0bfe93531eeb35fd870d18d4090a9d0))


### Bug Fixes

* remove unused var ([be69514](https://github.com/markusfalk/cd-config-server/commit/be69514e5eaca1acd432ee2cb39ec3baa5645673))

## [3.0.0](https://github.com/markusfalk/cd-config-server/compare/v2.2.0...v3.0.0) (2021-02-25)


### Features

* enable configs to be read from file system ([5b8a1d5](https://github.com/markusfalk/cd-config-server/commit/5b8a1d57e66e23cf6940a4ddd96404c95441aa3e))

## [2.2.0](https://github.com/markusfalk/cd-config-server/compare/v2.1.0...v2.2.0) (2021-01-30)


### Features

* enable cors with origin config ([c86f026](https://github.com/markusfalk/cd-config-server/commit/c86f0265cd901f4665850fc12192ee5f306a07fa)), closes [#2](https://github.com/markusfalk/cd-config-server/issues/2)

## [2.1.0](https://github.com/markusfalk/cd-config-server/compare/v2.0.0...v2.1.0) (2020-12-28)


### Features

* make gitlab repo url configurable ([9864430](https://github.com/markusfalk/cd-config-server/commit/986443087618f79eab6058412319b9accda01c91))
* rework homepage with link to issues ([fe1643a](https://github.com/markusfalk/cd-config-server/commit/fe1643a5a16a7365905bf4915f7ea89d02b94379))

## 2.0.0 (2020-12-23)

### BREAKING CHANGES

- add gitlab support

### Bug Fixes

- throw error on missing config values ([e5fa8f6](https://github.com/markusfalk/cd-config-server/commit/e5fa8f607df7a7b0a5200757ea1e964a31d2036f))

## 1.0.1 (2020-12-20)

- update documentation

## 1.0.0 (2020-12-19)

- Initial Release

### 0.0.2 (2020-12-10)

### Features

- add basic auth for rate limit ([6c9c38b](https://github.com/markusfalk/cd-config-server/commit/6c9c38b756d45b5e2e00d58e3686a511267533ab))
- add caching ([0e3c557](https://github.com/markusfalk/cd-config-server/commit/0e3c557470c2ee6e175c5ef5f173828d9b70d72f))
- add nest backend and load all tags from github ([27bfb73](https://github.com/markusfalk/cd-config-server/commit/27bfb737421e3b2a6373393b4b37f7c90d4b2005))
- add rate limit ([ec22ce2](https://github.com/markusfalk/cd-config-server/commit/ec22ce2c6aa1cae0950370c23992d7bc9cb8c292))
- connect to github ([905e77f](https://github.com/markusfalk/cd-config-server/commit/905e77f108c69e4cddfcf01971ff3322560422f7))
- dockerize server ([1b5556c](https://github.com/markusfalk/cd-config-server/commit/1b5556c3ed96363bbee14688b4a4ece87e4096ef))

### Bug Fixes

- add typings, remove logs ([39b0798](https://github.com/markusfalk/cd-config-server/commit/39b079877eda4d6d60c4635bb322fdacc6410f9f))
- only return latest config ([3b56b1f](https://github.com/markusfalk/cd-config-server/commit/3b56b1f3aa8e7064e39508183db8a685d95d23f8))
- reimport interfaces ([40ceb44](https://github.com/markusfalk/cd-config-server/commit/40ceb44eb270192f876a6cd4e51589a742a8dc37))
- remove env from build ([c08f76e](https://github.com/markusfalk/cd-config-server/commit/c08f76ea3220200268211ec098cd47183ea59313))
- remove hard coded github values ([9a7b3e3](https://github.com/markusfalk/cd-config-server/commit/9a7b3e3c604d64527e2d4443c0c9853426950eb5))


src/main.ts
import * as rateLimit from 'express-rate-limit';
import { join } from 'path';

import { CorsOptions } from '@nestjs/common/interfaces/external/cors-options.interface';
import { NestFactory } from '@nestjs/core';
import { NestExpressApplication } from '@nestjs/platform-express';
import { DocumentBuilder, SwaggerModule } from '@nestjs/swagger';

import { version } from './configuration/configuration/version.json';
import { AppModule } from './app.endpoint/app.module';
import * as helmet from 'helmet';

async function bootstrap() {
  const app = await NestFactory.create<NestExpressApplication>(AppModule);

  // Helmet helps you secure your Express apps by setting various HTTP headers. It's not a silver bullet, but it can help!
  app.use(helmet());

  // Swagger
  const options = new DocumentBuilder()
    .setTitle('Continuous Delivery Config Server')
    .setDescription(
      'Used to configure your app at runtime at different stages of your continuous delivery pipeline.',
    )
    .setVersion(version)
    // .addTag('configs') // TODO: find out what this is for
    .build();

  const document = SwaggerModule.createDocument(app, options);
  SwaggerModule.setup('api', app, document);

  // Rate Limit
  app.use(
    rateLimit({
      windowMs: parseFloat(process.env['RATE_LIMIT_MS']) || 360000,
      max: parseFloat(process.env['RATE_LIMIT_MAX']) || 5000,
    }),
  );

  // MVC
  app.useStaticAssets(join(__dirname, '..', 'src', '_assets'));
  app.setBaseViewsDir(join(__dirname, '..', 'src', '_views'));
  app.setViewEngine('hbs');

  // CORS
  const corsOptions: CorsOptions = {
    origin: process.env['CORS_ORIGIN'] || '*',
  };
  app.enableCors(corsOptions);

  // start app
  await app.listen(3000);
}

bootstrap();
